{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "\n",
    "def get_all_combinations(parent): # Finds all possible combinations of alleles a parent can pass on to their offspring, assuming independen assortment.\n",
    "\tif len(parent) == 1:\n",
    "\t\treturn [parent[0][0], parent[0][1]]\n",
    "\telse:\n",
    "\t\tgenlist = []\n",
    "\t\tfor x in get_all_combinations(parent[1:]):\n",
    "\t\t\tgenlist.append(parent[0][0] + x)\n",
    "\t\t\tgenlist.append(parent[0][1] + x)\n",
    "\t\treturn genlist\n",
    "\n",
    "def make_row(genotype, allele):\n",
    "\trow = []\n",
    "\tfor a in genotype:\n",
    "\t\trow.append(a + allele)\n",
    "\treturn row\n",
    "\n",
    "def make_table(parent1, parent2):\n",
    "\ttable = []\n",
    "\tfor a in parent1:\n",
    "\t\ttable.append(make_row(parent2, a))\n",
    "\treturn table\n",
    "\n",
    "def print_table(table, c1, c2): # formats and prints Punnett square\n",
    "\tlatextable = []\n",
    "\tdivlength = (len(c1[0])*2+4)*2**(len(c1[0]))\n",
    "\tprint('')\n",
    "\tprint('', end=' ')\n",
    "\tfor a in c2:\n",
    "\t\tprint(' '*(len(c1[0])+3) + a + '', end=' ')\n",
    "\t\tlatextable.append('& ' + a + ' ')\n",
    "\tprint('\\n' + ' '*(len(c1[0])+1) + '-'*(divlength))\n",
    "\tlatextable.append('\\\\\\ \\n\\\\hline\\n')\n",
    "\t\n",
    "\tfor i, row in enumerate(table):\n",
    "\t\tprint(c1[table.index(row)], end=' ')\n",
    "\t\tlatextable.append(c1[table.index(row)] + ' & ')\n",
    "\t\tprint('|', end=' ')\n",
    "\t\tfor j, cell in enumerate(row):\n",
    "\t\t\tprint(cell + ' | ', end=' ')\n",
    "\t\t\tif j != len(row)-1:\n",
    "\t\t\t\tlatextable.append(cell + ' & ')\n",
    "\t\t\telse:\n",
    "\t\t\t\tlatextable.append(cell + ' ')\n",
    "\t\tprint('\\n' + ' '*(len(c1[0])+1) + '-'*(divlength))\n",
    "\t\tif i != len(table)-1:\n",
    "\t\t\tlatextable.append('\\\\\\ \\n')\t\n",
    "\treturn latextable\t\t\n",
    "\t\n",
    "def print_genotype_frequencies(table): # calculates frequencies for each genotype present in table\n",
    "\tfreqtable = []\n",
    "\tfreqtable.append('\\n')\n",
    "\tcalculated = []\n",
    "\tgenotypes = [a for b in table for a in b]\n",
    "\tfor k, x in enumerate(genotypes):\n",
    "\t\tcount = 0\n",
    "\t\tfor y in genotypes:\n",
    "\t\t\tif sorted(x) == sorted(y):\n",
    "\t\t\t\tcount += 1\n",
    "\t\tif sorted(x) not in calculated:\n",
    "\t\t\tprint(\"The frequency of the \" + x + \" genotype is \" + str(float(count)/float((len(genotypes)))*100) + \"%.\")\n",
    "\t\t\tfreqtable.append(x + ' & ' + str(float(count)/float((len(genotypes)))*100) + '\\\\% \\\\\\ \\\\hline \\n')\t\n",
    "\t\tcalculated.append(sorted(x))\n",
    "\treturn freqtable\t\n",
    "\n",
    "\n",
    "def coin_toss(p):\n",
    "    '''\n",
    "    input: Probability to get a head \n",
    "    output: 1 for heads, 0 for tails \n",
    "    '''\n",
    "    coin = np.random.rand(1)\n",
    "    return np.sum(coin<p)\n",
    "def virtual_plant (N, traits, p=0.5, crosslinking = False, crosslinking_p = 0.5):\n",
    "    '''\n",
    "    input: N: The total number of plants to be simulated \n",
    "           traits: How many traits to be simulated \n",
    "           p: population propotion with dominace allele\n",
    "    output: List of dicts: \n",
    "            Each dict is in format: {0:[[0,1],[0]], 1:[[0,0],[0]],2:[[1,1],[0]]}\n",
    "            keys: Trait number \n",
    "            dict[key][0]: Two allelles, 0 means recessive, 1 means dominance \n",
    "            dict[key][1]: crosslinking marker, if it is 1, means there is cross linking, 0 means no cross linking\n",
    "    '''\n",
    "    plants = []\n",
    "    for i in range(N):\n",
    "        plant = {}\n",
    "        \n",
    "        for j in range(traits):\n",
    "            if crosslinking:\n",
    "                plant[j] = [[coin_toss(p),coin_toss(p)],[coin_toss(crosslinking_p)]]\n",
    "            else: \n",
    "                plant[j] = [[coin_toss(p),coin_toss(p)],[0]]\n",
    "        plants.append(plant)\n",
    "    return plants \n",
    "# Note of indexing: \n",
    "# plants[i] = plant \n",
    "# plant[j] /plants [i][j]= trait [j]\n",
    "# plant[j][0][0]/plants[i][j][0][0] = trait[i] allele 1, 1 is domin, 0 is recessive \n",
    "# plant[j][0][1]/plants[i][j][0][1]  = train[i] allele 2, 1 is domin, 0 is recessive\n",
    "# plant[j][1][0]/plants[i][j][1][0]  = crosslinking indicator, 1 is crosslink, 0 is no cross link. \n",
    "\n",
    "def cross_breeder(plant1, plant2, N=1, p =0.5):\n",
    "    '''\n",
    "    input: Plant 1: dict form, withe keys = trait index, [[allele1, allelle2], [crosslinking indicator]]\n",
    "           plant 2: same form as plant 1 \n",
    "           N: Cross breeding number, how much next generation generate\n",
    "    output: list of plants\n",
    "    '''\n",
    "    new_plants = []\n",
    "#     if plant1.keys().issubset(plant2.keys()) or plant2.keys().issubset(plant1.keys()):\n",
    "#         traits = np.min(len(plant1.keys(),len(plant2.keys())))\n",
    "#         ### These two lines makes sure: 1 list must have more traits than the other, and use the shorter traits. \n",
    "#         ### I am just showing I could do that, but there seems no need to do it. \n",
    "    for i in range(N):\n",
    "        new_plant = {}\n",
    "        crosslinker1 = coin_toss(p)\n",
    "        crosslinker2 = coin_toss(p)\n",
    "        allele1 = 2\n",
    "        allele2 = 2\n",
    "        for j in range(len(plant1.keys())):\n",
    "            if plant1[j][1][0]:\n",
    "                allele1 = plant1[j][0][crosslinker1]\n",
    "            else: \n",
    "                allele1 = plant1[j][0][coin_toss(p)]\n",
    "            if plant2[j][1][0]:\n",
    "                allele2 = plant2[j][0][crosslinker2]\n",
    "            else: \n",
    "                allele2 = plant2[j][0][coin_toss(p)]\n",
    "            new_plant[j] = [[allele1,allele2],[plant1[j][1][0]*plant2[j][1][0]]]\n",
    "        new_plants.append(new_plant)\n",
    "        return new_plants\n",
    "    \n",
    "alphabet0 = 'abcdefghijklmnopqrstuvwxyz'\n",
    "alphabet1 = alphabet0.upper()\n",
    "def genotype_converter(plant): \n",
    "    '''\n",
    "    input: a plant, with at most 26 traits\n",
    "    output: a string contains the alphabetical genotype, with A/a for the first allelle, and B/b and so on\n",
    "    \n",
    "    '''\n",
    "    string = []\n",
    "    for j in range(len(plant.keys())):\n",
    "        trait = sorted(plant[j][0])\n",
    "        if (trait[1]+trait[0])==2:\n",
    "            string.append(alphabet1[j] + alphabet1[j])\n",
    "        elif (trait[1] + trait[0])==1 :  \n",
    "            string.append(alphabet1[j] + alphabet0[j])\n",
    "        else: \n",
    "            string.append(alphabet0[j] + alphabet0[j])\n",
    "    return string\n",
    "def phenotype_converter(plant, method = 'auto'): \n",
    "    '''\n",
    "    input: a plant, with at most 26 traits\n",
    "    type: auto, or reversed \n",
    "    output: a string contains the alphabetical genotype, with A/a for the first allelle, and B/b and so on\n",
    "    \n",
    "    '''\n",
    "    phenotype = {}\n",
    "    if method == 'auto' :\n",
    "        string1 = 'Dominant'\n",
    "        string2 = 'Recessive'\n",
    "        string = ''\n",
    "        for j in range(len(plant.keys())):\n",
    "            string = alphabet1[j]\n",
    "            trait = sorted(plant[j][0])\n",
    "            if trait[1] or trait[0]:\n",
    "                phenotype[string] = string1\n",
    "            else: \n",
    "                phenotype[string] = string2\n",
    "    else:\n",
    "          \n",
    "        string1 = 'Dominant'\n",
    "        string2 = 'Recessive'\n",
    "        string = ''\n",
    "        for j in range(len(plant.keys())):\n",
    "            string = alphabet1[j]\n",
    "            trait = sorted(plant[j][0])\n",
    "            if trait[1] or trait[0]:\n",
    "                phenotype[string] = string2\n",
    "            else: \n",
    "                phenotype[string] = string1\n",
    "    return phenotype\n",
    "\n",
    "def reverse_genotype(string,crosslinking = False): \n",
    "    '''\n",
    "    input: string with 2 characters, like Aa/Bb or AA aa\n",
    "    output: a plant form: {x:[1,0],[0]}\n",
    "    '''\n",
    "    key = alphabet1.index(string[0].upper())\n",
    "    if string[0].isupper() and string[1].isupper():\n",
    "        genotype = [1,1]\n",
    "    elif string[0].islower() and string[1].islower():\n",
    "        genotype = [0,0]\n",
    "    else: \n",
    "        genotype = [0,1]\n",
    "    crosslinking_stat = np.sum(crosslinking)\n",
    "    plant = {key:[genotype,[crosslinking_stat]]}\n",
    "    return plant\n",
    "\n",
    "def punnett_square_generator(p1,p2):\n",
    "    ''' input: two Virtue plants \n",
    "        output: punnett_square print out \n",
    "    '''\n",
    "    p1string = genotype_converter(p1)\n",
    "    p2string = genotype_converter(p2)\n",
    "    c1 = get_all_combinations(p1string)\n",
    "    c2 = get_all_combinations(p2string)\n",
    "    a = make_table(c1, c2)\n",
    "    latextable = print_table(a, c1, c2)\n",
    "    freqtable = print_genotype_frequencies(a)\n",
    "    print('')\n",
    "    return latextable,freqtable\n",
    "\n",
    "# Plant class: Input a good genotype. Like AaBb. CcDd also works. Dont do it like #$%S>APD\n",
    "# crosslinking: [any genes are cross linked, like [1,2]]. Automatically, 1 should be B and 2 should be C. \n",
    "# you could do it sequntially, but I dont like it. \n",
    "# Plant.genotype: genotype of plant\n",
    "# Plant.phenotype: phenotype of plant \n",
    "# Plant.virtue: generate a virtue form of plant. This is the form from what we had later. \n",
    "# plant.cross: if takes no input, you get self cross. If you take another plant, you get cross breeding. \n",
    "# plant.punnett_square: get a punnett_square. \n",
    "\n",
    "\n",
    "class Plant:\n",
    "    def __init__(self, genotype, crosslinking = []):\n",
    "        self.genotype = genotype    # instance variable unique to each instance\n",
    "        virtue = {}\n",
    "        for i in range(int(len(genotype)/2)):\n",
    "            key = alphabet1.index(genotype[2*i].upper())\n",
    "            if genotype[2*i].isupper() and genotype[2*i+1].isupper():\n",
    "                vgenotype = [1,1]\n",
    "            elif genotype[2*i].islower() and genotype[2*i+1].islower():\n",
    "                vgenotype = [0,0]\n",
    "            else: \n",
    "                vgenotype = [0,1]\n",
    "            virtue[key] = [vgenotype,[0]]\n",
    "        for item in crosslinking:\n",
    "            virtue[item][1][0] = 1 \n",
    "        self.virtue = virtue\n",
    "        phenotype = phenotype_converter(virtue)\n",
    "        self.phenotype = phenotype\n",
    "        \n",
    "        def cross(self, plant2 = self):\n",
    "            cross = cross_breeder(self.virtue,plant2.virtue)\n",
    "            print(genotype_converter(cross))\n",
    "        def punnett_square(self, plant2 = self):\n",
    "            # here we need a pandas. 1st row: all genes father can pass, and 1st column is all genes mother can pass. \n",
    "            # we assume self is father \n",
    "            # 1st row: \n",
    "            punnett_square = punnett_square_generator(self.virtue, plant2.virtue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
